compiler/rustc_expand/src/mbe/transcribe.rs

```diff
@@ -564,6 +564,37 @@ fn transcribe_metavar_expr<'a>(
         span
     };
     match *expr {
+        MetaVarExpr::Concat(ref elements) => {
+            let mut concatenated = String::new();
+            for element in elements.into_iter() {
+                let s = 'string: {
+                    if !element.is_var {
+                        break 'string element.ident.to_string();
+                    }
+                    let span = element.ident.span;
+                    let mrni = MacroRulesNormalizedIdent::new(element.ident);
+                    if let Some(nm) = lookup_cur_matched(mrni, interp, &repeats)
+                        && let MatchedNonterminal(nt) = nm
+                    {
+                        if let Nonterminal::NtIdent(nt_ident, _) = &nt.0 {
+                            nt_ident.to_string()
+                        } else {
+                            return Err(cx.dcx().struct_span_err(
+                                span,
+                                "`${concat(..)}` currently only accepts identifiers as parameters",
+                            ));
+                        }
+                    } else {
+                        element.ident.to_string()
+                    }
+                };
+                concatenated.push_str(&s);
+            }
+            result.push(TokenTree::Token(
+                Token::from_ast_ident(Ident::new(Symbol::intern(&concatenated), visited_span())),
+                Spacing::Alone,
+            ));
```

Comment:
Looking at the tests I still don't quite get how hygiene works...`tests/ui/macros/rfc-3086-metavar-expr/concat-hygiene.rs` implies that `xy` and `${concat($x, $y)}` are different, but then `tests/ui/macros/rfc-3086-metavar-expr/concat.rs` uses things defined inside the macro outside of it...

Reply:
`Marker` is the responsible entity that process hygiene for tokens emitted by meta-variable expressions.

Its internal aspects are not entirely clear to me but if you want to know how everything works I guess @petrochenkov can give you a better answer.

Reply:
The current implementation marks the result as coming from the macro (using span of the metavar expression itself), regardless of contexts of the concatenated identifiers.
`concat_idents!()` macro works in a similar way, ignoring input contexts, but the output context is assigned in a different way.

What the desired behavior should be is a design question.
- Are we supposed to use local variables produced by concat inside `macro_rules` (macros 1.0) outside of the macro? Not sure.
- Are we supposed to use any names produced by concat inside `macro` (macros 2.0) outside of the macro?
  Likely yes, so some kind of hygiene opt out, or support for inheriting context from some token, should be added, but that applies to many things produced by `macro`, not just `concat`.

Reply:
@c410-f3r 
Could you copy this design question into the tracking issue as well?

Reply:
There is https://github.com/rust-lang/rust/issues/124225 which does link that question